File: src/services/ban.ts
// Ban 建议：无参数时按 ban 率/胜率推荐版本强势英雄；有我方阵容时推荐克制我方最多的
import { getChampionRankings, getHeroList, resolveHeroes, type RankOptions } from '../api/cn101.js';
import { heroDisplayName, type BanRecommendation, type ChampionBase, type ChampionStat } from '../models.js';

export interface BanInput {
  /** 我方已选英雄（可选），二选一 */
  myHeroes?: string[];
  myHeroIds?: number[];
  topN?: number;
  opts?: RankOptions;
}

export async function recommendBan(input: BanInput = {}): Promise<BanRecommendation[]> {
  const topN = input.topN ?? 10;
  const [heroes, rankings] = await Promise.all([
    getHeroList(),
    getChampionRankings(input.opts),
  ]);
  const myIds = input.myHeroIds
    ? new Set(input.myHeroIds)
    : input.myHeroes?.length
      ? new Set((await resolveHeroes(input.myHeroes)).map((r) => r.heroId))
      : null;
  return computeBanRecommendations(rankings, heroes, myIds, topN);
}

/**
 * 纯计算：不依赖网络，便于单测。
 * myIds 为 null/空 → 场景 A（版本强势榜）；否则 → 场景 B（克制我方最多的）。
 */
export function computeBanRecommendations(
  rankings: ChampionStat[],
  heroes: Map<number, ChampionBase>,
  myIds: Set<number> | null,
  topN: number,
): BanRecommendation[] {
  const byId = new Map(rankings.map((r) => [r.heroId, r]));

  // 场景 A：无我方阵容 —— 综合 ban 率 + 胜率 + 登场率（版本毒瘤榜）
  if (!myIds || myIds.size === 0) {
    const best = new Map<number, BanRecommendation>(); // 同一英雄多位置时保留分数最高的一条
    for (const r of rankings) {
      const rec: BanRecommendation = {
        heroId: r.heroId,
        title: heroDisplayName(heroes.get(r.heroId), r.heroId),
        winRate: r.winRate,
        pickRate: r.pickRate,
        banRate: r.banRate,
        tier: r.tier,
        threatensCount: 0,
        threatens: [],
        score: r.banRate * 1.0 + r.winRate * 0.5 + r.pickRate * 0.2,
      };
      const prev = best.get(r.heroId);
      if (!prev || rec.score > prev.score) best.set(r.heroId, rec);
    }
    return [...best.values()].sort((a, b) => b.score - a.score).slice(0, topN);
  }

  // 场景 B：有我方阵容 —— 找克制我方最多的英雄
  const myTitles = new Map([...myIds].map((id) => [id, heroDisplayName(heroes.get(id), id)]));
  const threatens = new Map<number, Set<number>>(); // 候选英雄 -> 被它克制(威胁)的我方英雄
  for (const stat of rankings) {
    if (myIds.has(stat.heroId)) continue; // 自己人不用 ban
    for (const c of stat.counters) {
      if (!myIds.has(c)) continue;
      if (!threatens.has(stat.heroId)) threatens.set(stat.heroId, new Set());
      threatens.get(stat.heroId)!.add(c);
    }
  }

  const recs: BanRecommendation[] = [];
  for (const [heroId, myCountered] of threatens) {
    const stat = byId.get(heroId);
    if (!stat) continue;
    recs.push({
      heroId,
      title: heroDisplayName(heroes.get(heroId), heroId),
      winRate: stat.winRate,
      pickRate: stat.pickRate,
      banRate: stat.banRate,
      tier: stat.tier,
      threatensCount: myCountered.size,
      threatens: [...myCountered].map((id) => myTitles.get(id) ?? String(id)),
      score: myCountered.size * 10 + stat.winRate + stat.banRate * 0.3,
    });
  }

  recs.sort((a, b) => b.score - a.score);
  return recs.slice(0, topN);
}

File: desktop/src-tauri/src/ban.rs
// Ban 建议（对应 src/services/ban.ts）：
// 场景 A（无我方阵容）：版本梯度榜 T0/T1 优先，组内按禁用率+胜率排序
// 场景 B（有我方阵容）：对位威胁分析（confront low_op = 被谁克制），按（威胁数、对位胜率、禁用率）综合排序

use crate::cn101::cn;
use crate::models::*;
use std::collections::{HashMap, HashSet};

pub struct BanInput {
    pub my_hero_names: Vec<String>,
    pub my_hero_ids: Vec<i64>,
    pub top_n: i64,
    pub tier: TierId,
}

pub async fn recommend_ban(input: &BanInput) -> Result<Vec<BanRecommendation>, String> {
    let top_n = input.top_n;
    let (heroes, rankings) = tokio::join!(
        cn().get_hero_list(false),
        cn().get_champion_rankings(input.tier, "ALL", None, false),
    );
    let heroes = heroes?;
    let rankings = rankings?;

    let my_ids: Option<HashSet<i64>> = if !input.my_hero_ids.is_empty() {
        Some(input.my_hero_ids.iter().copied().collect())
    } else if !input.my_hero_names.is_empty() {
        let resolved = cn().resolve_heroes(&input.my_hero_names).await?;
        Some(resolved.into_iter().map(|(id, _)| id).collect())
    } else {
        None
    };

    // 场景 B：对每个我方英雄查对位（被谁克制 = 威胁）
    let mut my_threats: Vec<(i64, i64, f64)> = Vec::new(); // (myId, threatHeroId, winRate)
    if let Some(ids) = &my_ids {
        if !ids.is_empty() {
            for my_id in ids {
                let lane = infer_lane_for(&rankings, *my_id);
                if let Ok(confront) = cn().get_confront(*my_id, input.tier, &lane, None, false).await {
                    for h in confront.low {
                        my_threats.push((*my_id, h.hero_id, h.win_rate));
                    }
                }
            }
        }
    }

    Ok(compute_ban_recommendations(&heroes, &rankings, my_ids, &my_threats, top_n))
}

fn infer_lane_for(rankings: &[ChampionStat], hero_id: i64) -> String {
    crate::pick::infer_lane(rankings, hero_id)
}

/// 纯计算：不依赖网络
pub fn compute_ban_recommendations(
    heroes: &HashMap<i64, ChampionBase>,
    rankings: &[ChampionStat],
    my_ids: Option<HashSet<i64>>,
    my_threats: &[(i64, i64, f64)],
    top_n: i64,
) -> Vec<BanRecommendation> {
    let tier_order = |t: &str| match t {
        "T0" => 0, "T1" => 1, "T2" => 2, "T3" => 3, _ => 4,
    };

    let mut my_ids = my_ids;
    // 有阵容但无威胁数据：降级为版本梯度榜
    if let Some(ids) = &my_ids {
        if !ids.is_empty() && my_threats.is_empty() {
            my_ids = None;
        }
    }

    // 场景 A：版本梯度榜
    if my_ids.is_none() || my_ids.as_ref().map(|s| s.is_empty()).unwrap_or(true) {
        let mut best: HashMap<i64, BanRecommendation> = HashMap::new();
        for r in rankings {
            let tier = tier_order(&r.tier);
            let rec = BanRecommendation {
                hero_id: r.hero_id,
                title: hero_display_name(heroes.get(&r.hero_id), r.hero_id),
                alias: heroes.get(&r.hero_id).map(|h| h.alias.clone()).unwrap_or_default(),
                win_rate: r.win_rate,
                pick_rate: r.pick_rate,
                ban_rate: r.ban_rate,
                tier: r.tier.clone(),
                lane: r.lane.clone(),
                threatens_count: 0,
                threatens: Vec::new(),
                matchups: Vec::new(),
                score: -tier as f64 * 1000.0 + r.ban_rate + r.win_rate * 0.5 + r.pick_rate * 0.2,
            };
            let prev = best.get(&r.hero_id);
            if prev.is_none() || rec.score > prev.unwrap().score {
                best.insert(r.hero_id, rec);
            }
        }
        let mut list: Vec<BanRecommendation> = best.into_values().collect();
        list.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
        list.truncate(top_n.max(0) as usize);
        return list;
    }

    // 场景 B：威胁并集
    let ids = my_ids.as_ref().unwrap();
    let mut candidates: HashMap<i64, Vec<(i64, f64)>> = HashMap::new();
    for (my_id, hero_id, win_rate) in my_threats {
        if ids.contains(hero_id) {
            continue; // 自己人不用 ban
        }
        candidates.entry(*hero_id).or_default().push((*my_id, *win_rate));
    }
    let by_id: HashMap<i64, &ChampionStat> = rankings.iter().map(|r| (r.hero_id, r)).collect();
    let my_title = |id: i64| hero_display_name(heroes.get(&id), id);

    let mut recs: Vec<BanRecommendation> = Vec::new();
    for (hero_id, threats) in &candidates {
        let Some(stat) = by_id.get(hero_id) else { continue };
        let avg = threats.iter().map(|t| t.1).sum::<f64>() / threats.len() as f64;
        let score = threats.len() as f64 * 10.0 + (avg - 50.0) * 0.8 + stat.ban_rate * 0.3 + stat.win_rate * 0.2;
        recs.push(BanRecommendation {
            hero_id: *hero_id,
            title: hero_display_name(heroes.get(hero_id), *hero_id),
            alias: heroes.get(hero_id).map(|h| h.alias.clone()).unwrap_or_default(),
            win_rate: stat.win_rate,
            pick_rate: stat.pick_rate,
            ban_rate: stat.ban_rate,
            tier: stat.tier.clone(),
            lane: stat.lane.clone(),
            threatens_count: threats.len() as i64,
            threatens: threats.iter().map(|t| my_title(t.0)).collect(),
            matchups: threats.iter().map(|t| BanMatchupRow {
                my_title: my_title(t.0),
                win_rate: t.1,
            }).collect(),
            score,
        });
    }
    recs.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));
    recs.truncate(top_n.max(0) as usize);
    recs
}
